### PR TITLE
fix(print-layout): exclude control colorbar canvases from map capture

### PR DESCRIPTION
**Pull request by giswqs**
## Problem

Opening the **Print Layout** with a raster on the map produced a preview where the entire map area was a smooth horizontal rainbow gradient instead of the actual map, even though the live map rendered correctly.

![bug: print preview shows a horizontal colormap ramp instead of the map]

## Root cause

`captureMapImage` (`apps/geolibre-desktop/src/lib/print-layout-export.ts`) composited the snapshot by drawing **every** `<canvas>` inside the map container, each stretched to fill the whole page:

```ts
const canvases = map.getContainer().querySelectorAll("canvas");
canvases.forEach((c) => ctx.drawImage(c, 0, 0, out.width, out.height));
```

The MapLibre raster control adds a small **colorbar/colormap preview** canvas to that same container, filled with a horizontal gradient (`_colorbarCanvas` … `createLinearGradient(0, 0, t, 0)` in `maplibre-gl-components`). That tiny canvas (~280×24) got stretched over the full A4 page and overwrote the real map with a rainbow ramp.

This is why the live map looked fine and the issue showed on the web build: on web the raster is interleaved into MapLibre's own base canvas, so the only stray full-stretch canvas is the colorbar preview.

## Fix

Composite only full-viewport render surfaces (the MapLibre base canvas and, on desktop, the deck.gl overlay), skipping the small control/preview canvases:

- **New** `apps/geolibre-desktop/src/lib/print-capture.ts` — dependency-free `isFullViewportMapCanvas(canvas, base)` helper (keeps a canvas only if it is the base, or ≥90% of the base width and height).
- `print-layout-export.ts` — `captureMapImage` early-returns for any canvas that fails that check.
- **New** `tests/print-capture.test.ts` — unit tests (base kept, full-size deck overlay kept, within-tolerance kept, small colorbar preview dropped, wide-but-short dropped, permissive when base size is unknown).

The helper lives in its own module so it stays unit-testable without pulling jsPDF / the Tauri filesystem bridge through `print-layout-export`.

## Testing

- `node --import tsx --test tests/print-capture.test.ts` — 6/6 pass.
- Existing `tests/print-layout.test.ts`, `tests/print-legend.test.ts`, `tests/layout-options.test.ts` — 37/37 pass.
- `pre-commit run --files …` — all hooks pass, including the full `npm build`.

## Notes

- Separate, pre-existing minor limitation (not addressed here): the print **legend** still shows a single neutral gray swatch for raster layers rather than the colormap ramp.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved map image export (PNG/PDF) to composite only the full-viewport map render canvas, preventing non-map canvases such as controls, previews/colormaps, and charts from overwriting or polluting the final snapshot.

* **Tests**
  * Added coverage for the canvas-selection logic, including identity handling for the base canvas and boundary checks for “near full-viewport” sizing and unknown base dimensions.

### CODE REVIEW

File: apps/geolibre-desktop/src/lib/print-capture.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * @returns True for the base canvas and any other canvas at least 90% of the
+ *   base's width and height; false for the smaller control/preview canvases.
+ */
+export function isFullViewportMapCanvas(
+  canvas: { width: number; height: number },
+  base: { width: number; height: number },
+): boolean {
+  if (canvas === base) return true;
+  // If the base size is unknown, be permissive rather than drop everything.
+  if (base.width === 0 || base.height === 0) return true;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/print-capture.ts]: Done in fb5882e — the 0×0 fallback now returns false for non-base canvases (`canvas === base` already returns true above), so a transiently-zeroed base can no longer let the colorbar canvas through. Updated the @param doc to note the identity check too.

File: tests/print-capture.test.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    );
+  });
+
+  it("keeps a canvas within the 90% size tolerance", () => {
+    assert.equal(
+      isFullViewportMapCanvas({ width: 1900, height: 1000 }, base),
+      true,
+    );
+  });
+
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/print-capture.test.ts:25]: Done in fb5882e — added a just-below-threshold case (1920×971 → false) that would fail if the 0.9 cutoff were mistyped, and updated the base-size-unknown test to assert base-kept / others-dropped.

File: tests/print-capture.test.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      false,
+    );
+  });
+
+  it("drops a canvas that is wide but short", () => {
+    assert.equal(
+      isFullViewportMapCanvas({ width: 1920, height: 24 }, base),
+      false,
+    );
+  });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/print-capture.test.ts:57]: Done in df8da11 — added a "tall but narrow" case (1727×1080 → false) mirroring the wide-but-short test, so an independent width-axis failure is now covered.

File: apps/geolibre-desktop/src/lib/print-capture.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+export function isFullViewportMapCanvas(
+  canvas: { width: number; height: number },
+  base: { width: number; height: number },
+): boolean {
+  if (canvas === base) return true;
+  // If the base size is unknown, keep only the base itself: compositing other
+  // canvases into a 0-sized output is a no-op anyway, and returning true here
+  // would let a control's colorbar canvas through and reintroduce the bug.
+  if (base.width === 0 || base.height === 0) return false;
+  return canvas.width >= base.width * 0.9 && canvas.height >= base.height * 0.9;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/print-capture.ts:45]: Done in df8da11 — documented in the JSDoc that this is a size heuristic (not an allow-list) and called out both edges: a sub-90% deck.gl overlay would be dropped, and a future near-full-screen widget canvas would be composited. Noted an explicit allow-list keyed on the overlay canvas as the fix if either case becomes real.

File: tests/print-capture.test.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  it("keeps a canvas within the 90% size tolerance", () => {
+    assert.equal(
+      isFullViewportMapCanvas({ width: 1900, height: 1000 }, base),
+      true,
+    );
+  });
+
+  it("drops a canvas just below the 90% threshold in height", () => {
+    // 1920 * 0.9 = 1728 (pass), 1080 * 0.9 = 972 -> 971 should fail.
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/print-capture.test.ts]: Done in df8da11 — reworded the comment to "1920 passes width; 971 < 972 fails height -> overall false", and added an exact-90%-boundary case (1728×972 → true) confirming the inclusive >= threshold.
